Clover/app/src/main/java/org/floens/chan/ui/captcha/v2/CaptchaNoJsHtmlParser.java

```diff
@@ -0,0 +1,398 @@
+        captchaInfo.setCheckboxes(new ArrayList<>(checkboxesSet));
+    }
+
+    @NonNull
+    private File downloadAndStoreImage(String fullUrl) throws IOException, CaptchaNoJsV2ParsingError {
+        Request request = new Request.Builder()
+                .url(fullUrl)
+                .build();
+
+        try (Response response = okHttpClient.newCall(request).execute()) {
+            if (response.code() != SUCCESS_STATUS_CODE) {
+                throw new CaptchaNoJsV2ParsingError(
+                        "Could not download challenge image, status code = " + response.code());
+            }
+
+            if (response.body() == null) {
+                throw new CaptchaNoJsV2ParsingError("Captcha challenge image body is null");
+            }
+
+            try (InputStream is = response.body().byteStream()) {
+                File imageFile = getChallengeImageFile();
+
+                try (OutputStream os = new FileOutputStream(imageFile)) {
+                    IOUtils.copy(is, os);
+                }
+
+                return imageFile;
+            }
+        }
+    }
+
+    private File getChallengeImageFile() throws CaptchaNoJsV2ParsingError, IOException {
+        File imageFile = new File(context.getCacheDir(), CHALLENGE_IMAGE_FILE_NAME);
+
+        if (imageFile.exists()) {
+            if (!imageFile.delete()) {
+                throw new CaptchaNoJsV2ParsingError(
+                        "Could not delete old challenge image file \"" + imageFile.getAbsolutePath() + "\"");
+            }
+        }
+
+        if (!imageFile.createNewFile()) {
+            throw new CaptchaNoJsV2ParsingError(
+                    "Could not create challenge image file \"" + imageFile.getAbsolutePath() + "\"");
+        }
+
+        if (!imageFile.canWrite()) {
+            throw new CaptchaNoJsV2ParsingError(
+                    "Cannot write to the challenge image file \"" + imageFile.getAbsolutePath() + "\"");
+        }
+
+        return imageFile;
+    }
+
+    private List<Bitmap> decodeImagesFromFile(File challengeImageFile, int columns, int rows) {
+        Bitmap originalBitmap = BitmapFactory.decodeFile(challengeImageFile.getAbsolutePath());
+        List<Bitmap> resultImages = new ArrayList<>(columns * rows);
+
+        try {
+            int imageWidth = originalBitmap.getWidth() / columns;
+            int imageHeight = originalBitmap.getHeight() / rows;
+
+            for (int column = 0; column < columns; ++column) {
+                for (int row = 0; row < rows; ++row) {
+                    Bitmap imagePiece = Bitmap.createBitmap(
+                            originalBitmap,
+                            column * imageWidth,
+                            row * imageHeight,
+                            imageWidth,
+                            imageHeight);
+
+                    resultImages.add(imagePiece);
+                }
+            }
+
+            return resultImages;
+        } finally {
```

Comment:
Should recycle bitmaps (in the catch clause) that were already added to the list when an exception occurs in the process of decoding the images. Right now they will leak.